lib/durations.nix

```diff
@@ -0,0 +1,245 @@
+/*
+ * Duration manipulation functions.
+ */
+
+{ lib }:
+
+with lib;
+with builtins;
+
+let
+
+  # Using the definitions of systemd.time(7).
+  unitToScale = rec {
+    us = 1;
+    usec = us;
+
+    ms = 1000 * us;
+    msec = ms;
+
+    s = 1000 * ms;
+    seconds = s;
+    second = s;
+    sec = s;
+
+    m = 60 * s;
+    minutes = m;
+    minute = m;
+    min = m;
+
+    h = 60 * m;
+    hours = h;
+    hour = h;
+    hr = h;
+
+    d = 24 * h;
+    days = d;
+    day = d;
+
+    w = 7 * d;
+    weeks = w;
+    week = w;
+
+    M = 2630016 * s;            # 30.44 days
+    months = M;
+    month = M;
+
+    y = 31557600 * s;           # 365.25 days
+    years = y;
+    year = y;
+  };
+
+  parseToNat = str:
+    let
+      maybeNat = fromJSON str;
+    in
+      if match "[[:space:]]*[[:digit:]]+" str != null && isInt maybeNat
+      then { ok = maybeNat; }
+      else { error = "Expected natural number but got \"${str}\"."; };
+
+  parseToScale = str:
+      if hasAttr str unitToScale
+      then { ok = unitToScale.${str}; }
+      else { error = "Expected time unit but got \"${str}\"."; };
+
+  scaleToMillis = magnitudeStr: unitStr:
+    let
+      maybeMagnitude = parseToNat magnitudeStr;
+      maybeScale = parseToScale unitStr;
+    in
+      if magnitudeStr == "" && unitStr == ""
+      then duration 0
+      else
+        if maybeMagnitude ? ok
+        then
+          if maybeScale ? ok
+          then duration (maybeMagnitude.ok * maybeScale.ok)
+          else maybeScale
+        else maybeMagnitude;
+
+  parseDuration = str:
+    let
+      err = msg: { error = "Invalid duration \"${str}\": ${msg}"; };
+      lstRaw = split "[[:space:]]*([a-zA-Z]+)[[:space:]]*" str;
+      lstParts = partition isString lstRaw;
+      magnitudes = lstParts.right;
+      units = map (xs: assert length xs == 1; head xs) lstParts.wrong ++ [""];
+      lstScaled = zipListsWith scaleToMillis magnitudes units;
+      add = acc: dur:
+        if acc ? duration
+        then
+          if dur ? duration
+          then duration (acc.duration + dur.duration)
+          else err dur.error
+        else acc;
+    in
+      if str == ""
+      then err "Cannot be empty string."
+      else foldl' add (duration 0) lstScaled;
+
+  # Duration value constructor.
+  duration = ms: {
```

Comment:
rename to `mkDuration` to indicate that it's a constructor?

Reply:
Nice idea, fixed.